lib/Resizable/index.tsx

```diff
@@ -0,0 +1,123 @@
+import type { PropsWithChildren } from "react";
+import React, { useCallback, useEffect, useRef } from "react";
+import { keepValueWithinRange, toPixels } from "../helpers";
+
+export interface ResizableProps {
+  defaultWidth?: number | string;
+  minWidth?: number | string;
+  maxWidth?: number | string;
+  useGutterOffset?: boolean;
+  containerWidth?: number | string;
+}
+
+/**
+ * A wrapper for making a given child element resizable
+ * @param children The element to be resized
+ * @param defaultWidth The width the child should initially render to. If min and max widths are provided, then the provided width will be massaged to meet those constraints
+ * @param maxWidth The maximum width the element can be resized to
+ * @param minWidth The minimum width the element can be resized to
+ * @param useGutterOffset Toggle whether we should take the width of the gutter into account for calculating the width of the resized item. This can be handy depending on the CSS being applied to or around the child element
+ * @param rest Contains params that need converting such as the containerWidth
+ */
+export function Resizable({
+  children,
+  defaultWidth = "50%",
+  maxWidth,
+  minWidth,
+  useGutterOffset = false,
+  ...rest
+}: PropsWithChildren<ResizableProps>) {
+  const containerWidth = toPixels(
+    rest.containerWidth ?? document.body.offsetWidth
+  );
+  const resizeRef = useRef<HTMLDivElement>(null);
+  const resizeWrapperRef = useRef<HTMLDivElement>(null);
+  const handleRef = useRef<HTMLSpanElement>(null);
+  const gutterSize = useGutterOffset ? 16 : 0;
+  const min = toPixels(minWidth ?? 0, containerWidth);
+  const max = toPixels(maxWidth ?? "100%", containerWidth);
+
+  const [state, setState] = React.useState({
+    startX: 0,
+    startWidth: 0,
+  });
+
+  const getWidth = () => toPixels(resizeWrapperRef.current?.style.width || 0);
+
+  const setWidth = (value: number) => {
+    if (resizeWrapperRef.current === null) return;
+
+    resizeWrapperRef.current.style.width = `${
+      keepValueWithinRange(value, min, max) - gutterSize
+    }px`;
+  };
+
+  const doDrag = (evt: MouseEvent) => {
+    const newWidth = state.startWidth + evt.clientX - state.startX;
+    setWidth(newWidth);
+  };
+
+  const stopDrag = () => {
+    document.removeEventListener("mousemove", doDrag, false);
+    document.removeEventListener("mouseup", stopDrag, false);
```

Comment:
These events will never be removed.

This is because on re-render `stopDrag` and `doDrag` will have a different signature and `removeEventListener` won't work.

Talk to @micmcgrorty about this, we had a similar issue with editor server events recently.

Reply:
Interesting point! I'll ask Michael about it :)